From 6d981bb836d603e6c0aa0fda000eead46a611589 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Thu, 18 Dec 2014 12:41:16 +0800
Subject: [PATCH 0775/1594] MLK-9986-4 dma: imx-sdma: add new ecspi tx script

commit 6d76bdcf2097e4198217edf27363cf6ba2e6542a from
git://git.freescale.com/imx/linux-2.6-imx.git

Current ecspi rom script didn't take care of rxfifo overflow risk. Add new
ecspi tx script to check the rxfifo status, if it is near to full(>=48 bytes),
do not copy data to txfifo which will trigger data push into rxfifo. Because
rx script may not read rxfifo in time, we have to consider it.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 17f472aa698aba0af5da4566df447e23306f4289)
(cherry picked from commit 90c929d7d1a3f8e196641b5ed7a33d2ee03bd63c)
---
 drivers/dma/imx-sdma.c                |    5 +-
 firmware/imx/sdma/sdma-imx6q.bin.ihex |  258 ++++++++++++++++-----------------
 2 files changed, 126 insertions(+), 137 deletions(-)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index 4d89b1f..91ac13f 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -7,7 +7,7 @@
  *
  * Based on code from Freescale:
  *
- * Copyright 2004-2014 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright 2004-2015 Freescale Semiconductor, Inc. All Rights Reserved.
  *
  * The code contained herein is licensed under the GNU General Public
  * License. You may obtain a copy of the GNU General Public License
@@ -796,6 +796,9 @@ static void sdma_get_pc(struct sdma_channel *sdmac,
 		emi_2_per = sdma->script_addrs->mcu_2_ata_addr;
 		break;
 	case IMX_DMATYPE_CSPI:
+		per_2_emi = sdma->script_addrs->app_2_mcu_addr;
+		emi_2_per = sdma->script_addrs->mcu_2_ecspi_addr;
+		break;
 	case IMX_DMATYPE_EXT:
 	case IMX_DMATYPE_SSI:
 	case IMX_DMATYPE_SAI:
diff --git a/firmware/imx/sdma/sdma-imx6q.bin.ihex b/firmware/imx/sdma/sdma-imx6q.bin.ihex
index abf10af..97a0908 100755
--- a/firmware/imx/sdma/sdma-imx6q.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx6q.bin.ihex
@@ -1,144 +1,130 @@
 :1000000053444D4103000000010000001C000000AB
-:1000100028000000BC000000280800008202000048
+:1000100029000000C0000000480700008202000024
 :10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFF6A1A00004B040000E5
-:10005000EB020000BB180000FFFFFFFFC81A000002
+:10004000FFFFFFFFFFFFFFFFFA1900004B04000056
+:10005000EB020000A8180000FFFFFFFF581A000085
 :10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
 :10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
 :10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
-:1000A000000000000018000062180000161A00008E
-:1000B000771B0000121B0000AE1B0000E3C1DB57E2
-:1000C000E35FE357F352016A8F00D500017D8D0095
-:1000D000A005EB5D7804037D79042C7D367C7904E2
-:1000E0001F7CEE56000F6006057D0965437E0A629F
-:1000F000417E20980A623E7E09653C7E120512050B
-:10010000AD026007037DFB55D36D2B98FB55041D95
-:10011000D36DC86A2F7F011F03200048E47C5398E9
-:10012000FB55D76D150005780962C86A0962C86A6F
-:10013000D76D5298FB55D76D1500150005780A62EA
-:10014000C86A0A62C86AD76D5298FB55D76D150008
-:100150001500150005780B62C86A0B62C86AD76D76
-:10016000097CDF6D077F0000EB55004D077DFAC16C
-:10017000E35706980700CC680C6813C20AC20398BC
-:10018000D9C1E3C1DB57E35FE357F352216A8F0024
-:10019000D500017D8D00A005EB5DFB567804037D45
-:1001A00079042A7D317C7904207C700B1103EB5398
-:1001B000000F6003057D0965377E0A62357E8698EB
-:1001C0000A62327E0965307E12051205AD026007B3
-:1001D000027C065A8E98265A277F011F032000486A
-:1001E000E87C700B11031353AF9815000478096273
-:1001F000065A0962265AAE981500150004780A625C
-:10020000065A0A62265AAE981500150015000478A1
-:100210000B62065A0B62265A077C0000EB55004D14
-:10022000067DFAC1E357699807000C6813C20AC239
-:100230006698700B110313536C07017CD9C1FB5EE8
-:100240008A066B07017CD9C1F35EDB59D3588F0155
-:1002500010010F398B003CC12B7DC05AC85B4EC1C9
-:10026000277C88038906E35CFF0D1105FF1DBC0593
-:100270003E07004D187D700811007E07097D7D073F
-:10028000027D2852E698F852DB54BC02CC02097C6D
-:100290007C07027D2852EF98F852D354BC02CC025E
-:1002A000097D0004DD988B00C052C85359C1D67D2A
-:1002B0000002CD98FF08BF007F07157D8804D50098
-:1002C000017D8D00A005EB5D8F0212021202FF3A44
-:1002D000DA05027C3E071899A402DD02027D3E0782
-:1002E00018995E071899EB559805EB5DF352FB548E
-:1002F0006A07267D6C07017D55996B07577C690756
-:10030000047D6807027D010E2F999358D600017D68
-:100310008E009355A005935DA00602780255045DFA
-:100320001D7C004E087C6907037D0255177E3C99B1
-:10033000045D147F890693500048017D2799A09998
-:10034000150006780255045D4F070255245D2F07FE
-:10035000017CA09917006F07017C012093559D0037
-:100360000700A7D9F598D36C6907047D6807027D5B
-:10037000010E64999358D600017D8E009355A00517
-:10038000935DA00602780255C86D0F7C004E087C74
-:100390006907037D0255097E7199C86D067F89063C
-:1003A00093500048017D5C99A0999A99C36A6907A6
-:1003B000047D6807027D010E87999358D600017D60
-:1003C0008E009355A005935DA0060278C865045D74
-:1003D0000F7C004E087C6907037DC865097E9499EF
-:1003E000045D067F890693500048017D7F99A0999E
-:1003F00093559D000700FF6CA7D9F5980000E354C2
-:10040000EB55004D017CF598DD98E354EB55FF0A60
-:100410001102FF1A7F07027CA005B4999D008C058C
-:10042000BA05A0051002BA04AD0454040600E3C1E5
-:10043000DB57FB52C36AF352056A8F00D500017D7A
-:100440008D00A005EB5D7804037D79042B7D1E7C77
-:100450007904337CEE56000FFB556007027DC36DB7
-:10046000D599041DC36DC8623B7E6006027D1002F3
-:100470001202096A357F1202096A327F1202096A82
-:100480002F7F011F03200048E77C099AFB55C76DA9
-:100490001500150015000578C8620B6AC8620B6A62
-:1004A000C76D089AFB55C76D150015000578C86221
-:1004B0000A6AC8620A6AC76D089AFB55C76D1500BB
-:1004C0000578C862096AC862096AC76D097C286A2A
-:1004D000077F0000EB55004D057DFAC1DB57BF9942
-:1004E00077C254040AC2BA99D9C1E3C1DB57F352A7
-:1004F000056A8F00D500017D8D00A005FB567804AC
-:10050000037D7904297D1F7C79042E7CE35D700DC9
-:100510001105ED55000F6007027D0652329A2652F2
-:10052000337E6005027D10021202096A2D7F1202DD
-:10053000096A2A7F1202096A277F011F03200048E7
-:10054000EA7CE3555D9A1500150015000478065203
-:100550000B6A26520B6A5C9A150015000478065245
-:100560000A6A26520A6A5C9A150004780652096AD9
-:100570002652096A097C286A077F0000DB57004D74
-:10058000057DFAC1DB571B9A77C254040AC2189A38
-:10059000E3C1DB57F352056AFB568E02941AC36A15
-:1005A000C8626902267D941EC36ED36EC86248027B
-:1005B000C86A9426981EC36EC8629826C36E6002ED
-:1005C0000E7D981EC36EC8626C02037D9826C36EB2
-:1005D000AB9A4C02D36EC86A9826C36EBB9AC862A7
-:1005E0006E02017CB59A096A1A7F0125004D217DB2
-:1005F0007C9AE36E8F00D805017D8D00C8626E0283
-:10060000127D096A0C7F01250120F87CDB57004D23
-:10061000107D286A047F0000FAC1DB576E9A07003C
-:1006200004620C6AB89A286AFA7F04627AC2580493
-:100630005404FF081100FF18BC00CD00017CAB9AE8
-:10064000286AED7F04627AC20AC26B9AD9C1E3C1FB
-:10065000DB57F352056AFB568E02941A0252690266
-:100660001D7D941E06524802065A9426981E065274
-:100670004C02065A9826981E065260020A7C98265A
-:1006800006526E02237D096A1D7F0125004D247DDF
-:10069000DD9A286A177F04627AC20E9B8F00D80504
-:1006A000017D8D00A00506526E02107D096A0A7F49
-:1006B0000120F97C286A067F0000004D0D7DFAC1FB
-:1006C000DB57CD9A070004620C6A0B9B286AFA7FFD
-:1006D00004627AC258045404286AF47F0AC2CA9A8F
-:1006E00056DBDB57F352056AC7698F00D500017DE1
-:1006F0008D00A0057804037D79041C7D157C7904A8
-:100700001E7CEE56C862287E6006027D1002120230
-:10071000096A227F1202096A1F7F1202096A1C7F7E
-:1007200003200048EF7C459B150015000278C86245
-:100730000B6A449B15000278C8620A6A449B0278DF
-:10074000C862096A097C286A077F004D077DD35279
-:10075000010802580004169B77C25404D3520108C2
-:10076000025801046EDB139B700B11031353DB5F04
-:100770000A07D3588B00FB5E3CC1157DC05AC85B8D
-:100780004EC1117C8803F05DFF0D1105FF1DBC05F6
-:10079000004D047D6D9B0807DB5F0A078B00C35289
-:1007A000CB5359C1EC7D00025E9B016E0B612F7E25
-:1007B0000B622D7E0B632B7E0C0D170417041704A0
-:1007C0009D04081DCC05017C0C0DD16A000F420769
-:1007D000C86FDD6F1C7F8E009D0001680B67177E60
-:1007E000D56B04080278C86F1207117C0B670F7E67
-:1007F00004080278C86F12070A7CDD6F087FD16990
-:10080000010FC86FDD6F037F01010004839B0700A8
-:10081000FF680C680002839BF3DBDB57F352056A29
-:10082000FB52C76A8F00D500017D8D00A0057804BA
-:10083000037D79041C7D157C79041E7CEE56C8620C
-:10084000287E6006027D10021202096A227F1202CF
-:10085000096A1F7F1202096A1C7F03200048EF7C8F
-:10086000E29B150015000278C8620B6AE19B150037
-:100870000278C8620A6AE19B0278C862096A097C48
-:10088000286A077F004D077DD352010802580004F3
-:10089000B29B77C25404D3520108025801040BDC06
-:1008A000AF9B700B11031353DB5F0A07D3588B0008
-:1008B000FB5E3CC1157DC05AC85B4EC1117C8803EC
-:1008C000F05DFF0D1105FF1DBC05004D047D0A9C68
-:1008D0000807DB5F0A078B00C352CB5359C1EC7D7D
-:0408E0000002FB9B7C
+:1000A00000000000001800004F180000A619000012
+:1000B000071B0000A21A00003E1B000000180000F1
+:1000C000E3C1DB57F352016AFB52D36AFB521C1A9D
+:1000D000C36AE8621102FF3A3008D002337C8F0015
+:1000E000D500017D8D00A005EB5D7804037D7904CA
+:1000F0001C7D207C7904157CEE566006057D096523
+:10010000287E0A62267E28980A62237E0965217E5F
+:1001100012051205AD02C86A1C7F03200048ED7C61
+:10012000409802780962C86A3F98150002780A620E
+:10013000C86A3F981500150002780B62C86A097CEE
+:10014000DF6D077F0000EB55004D077DFAC1DB57DF
+:1001500006980700CC680C6813C20AC20198D9C17E
+:10016000E3C1DB57E35FE357F352216A8F00D50009
+:10017000017D8D00A005EB5DFB567804037D7904BD
+:100180002A7D317C7904207C700B1103EB53000F26
+:100190006003057D0965377E0A62357E73980A62C1
+:1001A000327E0965307E12051205AD026007027CC1
+:1001B000065A7B98265A277F011F03200048E87CB7
+:1001C000700B110313539C98150004780962065AAA
+:1001D0000962265A9B981500150004780A62065A8F
+:1001E0000A62265A9B9815001500150004780B62C8
+:1001F000065A0B62265A077C0000EB55004D067D1F
+:10020000FAC1E357569807000C6813C20AC2539804
+:10021000700B110313536C07017CD9C1FB5E8A0676
+:100220006B07017CD9C1F35EDB59D3588F011001F4
+:100230000F398B003CC12B7DC05AC85B4EC1277C57
+:1002400088038906E35CFF0D1105FF1DBC053E0711
+:10025000004D187D700811007E07097D7D07027D25
+:100260002852D398F852DB54BC02CC02097C7C079C
+:10027000027D2852DC98F852D354BC02CC02097D8E
+:100280000004CA988B00C052C85359C1D67D0002E1
+:10029000BA98FF08BF007F07157D8804D500017D4F
+:1002A0008D00A005EB5D8F0212021202FF3ADA0503
+:1002B000027C3E070599A402DD02027D3E070599F6
+:1002C0005E070599EB559805EB5DF352FB546A0701
+:1002D000267D6C07017D42996B07577C6907047D79
+:1002E0006807027D010E1C999358D600017D8E008F
+:1002F0009355A005935DA00602780255045D1D7C10
+:10030000004E087C6907037D0255177E2999045D1C
+:10031000147F890693500048017D14998D9915002A
+:1003200006780255045D4F070255245D2F07017CB6
+:100330008D9917006F07017C012093559D000700E0
+:1003400094D9E298D36C6907047D6807027D010E99
+:1003500051999358D600017D8E009355A005935D69
+:10036000A00602780255C86D0F7C004E087C690714
+:10037000037D0255097E5E99C86D067F89069350FC
+:100380000048017D49998D998799C36A6907047D61
+:100390006807027D010E74999358D600017D8E0086
+:1003A0009355A005935DA0060278C865045D0F7C97
+:1003B000004E087C6907037DC865097E8199045D4C
+:1003C000067F890693500048017D6C998D9993555D
+:1003D0009D000700FF6C94D9E2980000E354EB55B0
+:1003E000004D017CE298CA98E354EB55FF0A1102D4
+:1003F000FF1A7F07027CA005A1999D008C05BA0514
+:10040000A0051002BA04AD0454040600D9C1E3C12A
+:10041000DB57F352056A8F00D500017D8D00A005E2
+:10042000FB567804037D7904297D1F7C79042E7C9A
+:10043000E35D700D1105ED55000F6007027D06525A
+:10044000C2992652337E6005027D10021202096AAB
+:100450002D7F1202096A2A7F1202096A277F011F73
+:1004600003200048EA7CE355ED99150015001500BE
+:10047000047806520B6A26520B6AEC991500150097
+:10048000047806520A6A26520A6AEC991500047822
+:100490000652096A2652096A097C286A077F000009
+:1004A000DB57004D057DFAC1DB57AB9977C2540489
+:1004B0000AC2A899E3C1DB57F352056AFB568E02C4
+:1004C000941AC36AC8626902267D941EC36ED36EF5
+:1004D000C8624802C86A9426981EC36EC8629826ED
+:1004E000C36E60020E7D981EC36EC8626C02037DEF
+:1004F0009826C36E3B9A4C02D36EC86A9826C36E88
+:100500004B9AC8626E02017C459A096A1A7F0125DE
+:10051000004D217D0C9AE36E8F00D805017D8D0082
+:10052000C8626E02127D096A0C7F01250120F87CE9
+:10053000DB57004D107D286A047F0000FAC1DB57AD
+:10054000FE99070004620C6A489A286AFA7F0462DE
+:100550007AC258045404FF081100FF18BC00CD00F3
+:10056000017C3B9A286AED7F04627AC20AC2FB9939
+:10057000D9C1E3C1DB57F352056AFB568E02941AC8
+:10058000025269021D7D941E06524802065A9426A4
+:10059000981E06524C02065A9826981E0652600271
+:1005A0000A7C982606526E02237D096A1D7F01256A
+:1005B000004D247D6D9A286A177F04627AC29E9A44
+:1005C0008F00D805017D8D00A00506526E02107DBA
+:1005D000096A0A7F0120F97C286A067F0000004D25
+:1005E0000D7DFAC1DB575D9A070004620C6A9B9A85
+:1005F000286AFA7F04627AC258045404286AF47F95
+:100600000AC25A9AE6DADB57F352056AC7698F00C5
+:10061000D500017D8D00A0057804037D79041C7D43
+:10062000157C79041E7CEE56C862287E6006027D29
+:1006300010021202096A227F1202096A1F7F120247
+:10064000096A1C7F03200048EF7CD59A150015002D
+:100650000278C8620B6AD49A15000278C8620A6AE6
+:10066000D49A0278C862096A097C286A077F004D1B
+:10067000077DD352010802580004A69A77C2540499
+:10068000D352010802580104FEDAA39A700B110339
+:100690001353DB5F0A07D3588B00FB5E3CC1157D0B
+:1006A000C05AC85B4EC1117C8803F05DFF0D110577
+:1006B000FF1DBC05004D047DFD9A0807DB5F0A079E
+:1006C0008B00C352CB5359C1EC7D0002EE9A016EF0
+:1006D0000B612F7E0B622D7E0B632B7E0C0D17049E
+:1006E000170417049D04081DCC05017C0C0DD16A6C
+:1006F000000F4207C86FDD6F1C7F8E009D000168F0
+:100700000B67177ED56B04080278C86F1207117C3F
+:100710000B670F7E04080278C86F12070A7CDD6F32
+:10072000087FD169010FC86FDD6F037F01010004ED
+:10073000139B0700FF680C680002139B83DBDB57E9
+:10074000F352056AFB52C76A8F00D500017D8D0008
+:10075000A0057804037D79041C7D157C79041E7C3A
+:10076000EE56C862287E6006027D10021202096AF7
+:10077000227F1202096A1F7F1202096A1C7F03206E
+:100780000048EF7C729B150015000278C8620B6A66
+:10079000719B15000278C8620A6A719B0278C86270
+:1007A000096A097C286A077F004D077DD35201083A
+:1007B00002580004429B77C25404D35201080258E5
+:1007C00001049BDB3F9B700B11031353DB5F0A0794
+:1007D000D3588B00FB5E3CC1157DC05AC85B4EC12F
+:1007E000117C8803F05DFF0D1105FF1DBC05004D58
+:1007F000047D9A9B0807DB5F0A078B00C352CB532B
+:0808000059C1EC7D00028B9B45
 :00000001FF
-- 
1.7.5.4

